packages/core-data/src/batch/create-batch.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * External dependencies
+ */
+import { isFunction, zip } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import defaultProcessor from './default-processor';
+
+/**
+ * Creates a batch, which can be used to combine multiple API requests into one
+ * API request using the WordPress batch processing API (/v1/batch).
+ *
+ * ```
+ * const batch = createBatch();
+ * const dunePromise = batch.add( {
+ *   path: '/v1/books',
+ *   method: 'POST',
+ *   data: { title: 'Dune' }
+ * } );
+ * const lotrPromise = batch.add( {
+ *   path: '/v1/books',
+ *   method: 'POST',
+ *   data: { title: 'Lord of the Rings' }
+ * } );
+ * const isSuccess = await batch.run(); // Sends one POST to /v1/batch.
+ * if ( isSuccess ) {
+ *   console.log(
+ *     'Saved two books:',
+ *     await dunePromise,
+ *     await lotrPromise
+ *   );
+ * }
+ * ```
+ *
+ * @param {Function} [processor] Processor function. Can be used to replace the
+ *                               default functionality which is to send an API
+ *                               request to /v1/batch. Is given an aray of
+ *                               inputs and must return a promise that
+ *                               resolves to an array of objects containing
+ *                               either `output` or `error`.
+ */
+export default function createBatch( processor = defaultProcessor ) {
+	const queue = [];
+
+	let expectedSize = 0,
+		actualSize = 0,
+		onFull = () => {};
+
+	return {
+		/**
+		 * Adds an input to the batch and returns a promise that is resolved or
+		 * rejected when the input is processed by `batch.run()`.
+		 *
+		 * You may also pass a thunk which allows inputs to be added
+		 * asychronously.
+		 *
+		 * ```
+		 * // Both are allowed:
+		 * batch.add( { path: '/v1/books', ... } );
+		 * batch.add( ( add ) => add( { path: '/v1/books', ... } ) );
+		 * ```
+		 *
+		 * @param {any|Function} inputOrThunk Input to add or thunk to execute.
+		 
+		 * @return {Promise|any} If given an input, returns a promise that
+		 *                       is resolved or rejected when the batch is
+		 *                       processed. If given a thunk, returns the return
+		 *                       value of that thunk.
+		 */
+		add( inputOrThunk ) {
+			++expectedSize;
+
+			const add = ( input ) =>
+				new Promise( ( resolve, reject ) => {
+					queue.push( {
+						input,
+						resolve,
+						reject,
+					} );
+
+					if ( ++actualSize === expectedSize ) {
+						onFull();
+					}
```

Comment:
Comparing the `actualSize` and `expectedSize` solves the pressing issue with saving widgets, but as a general-purpose utility, it makes some assumptions that don't need to be always true.

Namely, it assumes that each call like:
```js
batch.add( add => {
  dispatch( 'core' ).saveSomething( { __fetch: add } );
} );
```
does exactly one fetch and calls the `add` callback exactly once. If the `saveSomething` action does two fetches or no fetch at all, it will break -- either by waiting forever or stopping the wait too early.

`saveEditedEntityRecord` will skip the fetch if the record supplied doesn't have any edits, for example.

I don't know how to solve this better. Maybe after rewritten from generators to normal async functions, the code will be clearer and easier to reason about and some solutions will pop up.

Reply:
Good catch @jsnajdr. I've tweaked it so that `run` will unpause if `thunk` resolves or returns without calling `add()`. See 79d9829b2f.

Agree that switching to `async` functions will at least make this easier to follow.